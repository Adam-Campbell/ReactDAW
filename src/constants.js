export const synthTypes = {
    default: 'synth',
    am: 'amSynth',
    fm: 'fmSynth',
    duo: 'duoSynth',
    mono: 'monoSynth'
};

export const UIColors = {
    deepPurple: '#201826',
    lightPurple: '#47426c',
    pink: '#d86597',
    offWhite: '#e0e0e0',
    brightBlue: '#00bff3',
    brightYellow: '#f4ab54',
    brightRed: '#e23340',
    brightGreen: '#25e452'
};

export const effectTypes = {
    autoFilter: 'autoFilter',
    bitCrusher: 'bitCrusher',
    chorus: 'chorus',
    compressor: 'compressor',
    distortion: 'distortion',
    eq3: 'eq3',
    feedbackDelay: 'feedbackDelay',
    filter: 'filter',
    freeverb: 'freeverb',
    gate: 'gate',
    jcReverb: 'jcReverb',
    lfo: 'lfo',
    limiter: 'limiter',
    multibandCompressor: 'multibandCompressor',
    phaser: 'phaser',
    pingPongDelay: 'pingPongDelay',
    pitchShift: 'pitchShift',
    tremolo: 'tremolo',
    vibrato: 'vibrato'
};

export const defaultSynthData = {
    envelope: {
        attack: 0.005,
        attackCurve: 'linear',
        decay: 0.1,
        release: 1,
        releaseCurve: 'exponential',
        sustain: 0.3
    },
    oscillator: {
        detune: 0,
        mute: false,
        phase: 0,
        type: 'triangle',
        volume: 0
    },
    portamento: 0,
    volume: 0
}

export const synthData = {
    synth: {
        envelope: {
            attack: 0.005,
            attackCurve: 'linear',
            decay: 0.1,
            release: 1,
            releaseCurve: 'exponential',
            sustain: 0.3
        },
        oscillator: {
            detune: 0,
            mute: false,
            phase: 0,
            type: 'triangle',
            volume: 0
        },
        portamento: 0,
        volume: 0
    },
    amSynth: {
        detune: 0,
        envelope: {
            attack: 0.01,
            attackCurve: 'linear',
            decay: 0.01,
            release: 0.5,
            releaseCurve: 'exponential',
            sustain: 1
        },
        harmonicity: 3,
        modulation: {
            detune: 0,
            mute: false,
            phase: 0,
            type: 'square',
            volume: 0
        },
        modulationEnvelope: {
            attack: 0.5,
            attackCurve: 'linear',
            decay: 0,
            release: 0.5,
            releaseCurve: 'exponential',
            sustain: 1
        },
        oscillator: {
            detune: 0,
            mute: false,
            phase: 0,
            type: 'triangle',
            volume: 0
        },
        portamento: 0,
        volume: 0
    },
    fmSynth: {
        detune: 0,
        envelope: {
            attack: 0.01,
            attackCurve: 'linear',
            decay: 0.01,
            release: 0.5,
            releaseCurve: 'exponential',
            sustain: 1
        },
        harmonicity: 3,
        modulation: {
            detune: 0,
            mute: false,
            phase: 0,
            type: 'square',
            volume: 0
        },
        modulationEnvelope: {
            attack: 0.5,
            attackCurve: 'linear',
            decay: 0,
            release: 0.5,
            releaseCurve: 'exponential',
            sustain: 1
        },
        modulationIndex: 10,
        oscillator: {
            detune: 0,
            mute: false,
            phase: 0,
            type: 'triangle',
            volume: 0
        },
        portamento: 0,
        volume: 0
    },
    duoSynth: {
        harmonicity: 1.5,
        portamento: 0,
        vibratoAmount: 0.5,
        vibratoRate: 5,
        voice0: {
            detune: 0,
            envelope: {
                attack: 0.01,
                attackCurve: 'linear',
                decay: 0,
                release: 0.5,
                releaseCurve: 'exponential',
                sustain: 1
            },
            filter: {
                Q: 6,
                frequency: 350,
                gain: 0,
                rolloff: -24,
                type: 'lowpass'
            },
            filterEnvelope: {
                attack: 0.01,
                attackCurve: 'linear',
                baseFrequency: 200,
                decay: 0,
                exponent: 2,
                octaves: 7,
                release: 0.5,
                releaseCurve: 'exponential',
                sustain: 1
            },
            oscillator: {
                detune: 0,
                mute: false,
                phase: 0,
                type: 'sine',
                volume: 0
            },
            portamento: 0,
            volume: -9.999999999999998
        },
        voice1: {
            detune: 0,
            envelope: {
                attack: 0.01,
                attackCurve: 'linear',
                decay: 0,
                release: 0.5,
                releaseCurve: 'exponential',
                sustain: 1
            },
            filter: {
                Q: 6,
                frequency: 350,
                gain: 0,
                rolloff: -24,
                type: 'lowpass'
            },
            filterEnvelope: {
                attack: 0.01,
                attackCurve: 'linear',
                baseFrequency: 200,
                decay: 0,
                exponent: 2,
                octaves: 7,
                release: 0.5,
                releaseCurve: 'exponential',
                sustain: 1
            },
            oscillator: {
                detune: 0,
                mute: false,
                phase: 0,
                type: 'sine',
                volume: 0
            },
            portamento: 0,
            volume: -9.999999999999998
        },
        volume: 0
    },
    monoSynth: {
        detune: 0,
        envelope: {
            attack: 0.005,
            attackCurve: 'linear',
            decay: 0.1,
            release: 1,
            releaseCurve: 'exponential',
            sustain: 0.9
        },
        filter: {
            Q: 6,
            frequency: 350,
            gain: 0,
            rolloff: -24,
            type: 'lowpass'
        },
        filterEnvelope: {
            attack: 0.06,
            attackCurve: 'linear',
            baseFrequency: 200,
            decay: 0.2,
            exponent: 2,
            octaves: 7,
            release: 2,
            releaseCurve: 'exponential',
            sustain: 0.5
        },
        oscillator: {
            detune: 0,
            mute: false,
            phase: 0,
            type: 'square',
            volume: 0
        },
        portamento: 0,
        volume: 0
    }
};


export const effectData = {
    autoFilter: {
        baseFrequency: 200,
        depth: 1,
        filter: {
            Q: 1,
            gain: 0,
            rolloff: -12,
            type: 'lowpass'
        },
        frequency: 1,
        octaves: 2.6,
        type: 'sine',
        wet: 1
    },
    bitCrusher: {
        bits: 4,
        wet: 1
    },
    chorus: {
        delayTime: 3.5,
        depth: 0.7,
        frequency: 1.5,
        spread: 180,
        type: 'sine',
        wet: 1
    },
    compressor: {
        attack: 0.003,
        knee: 30,
        ratio: 12,
        release: 0.25,
        threshold: -24
    },
    distortion: {
        distortion: 0.4,
        oversample: 'none',
        wet: 1
    },
    eq3: {
        high: 0,
        highFrequency: 2500,
        low: 0,
        lowFrequency: 400,
        mid: 0
    },
    feedbackDelay: {
        delayTime: 0.25,
        feedback: 0.125,
        wet: 1
    },
    filter: {
        Q: 1,
        frequency: 350,
        gain: 0,
        rolloff: -12,
        type: 'lowpass'
    },
    freeverb: {
        dampening: 3000,
        roomSize: 0.7,
        wet: 1
    },
    gate: {
        attack: 0.1, 
        release: 0.1,
        threshold: -39.99999999999999
    },
    jcReverb: {
        roomSize: 0.5,
        wet: 1,
    },
    lfo: {
        amplitude: 1,
        frequency: 2,
        max: 1,
        min: 0,
        phase: 0,
        type: 'sine',
        units: 'number'
    },
    limiter: {
        threshold: -12
    },
    multibandCompressor: {
        high: {
            attack: 0.003,
            knee: 30,
            ratio: 12,
            release: 0.25,
            threshold: -24
        },
        highFrequency: 2500,
        low: {
            attack: 0.003,
            knee: 30,
            ratio: 12,
            release: 0.25,
            threshold: -24
        },
        lowFrequency: 400,
        mid: {
            attack: 0.003,
            knee: 30,
            ratio: 12,
            release: 0.25,
            threshold: -24
        }
    },
    phaser: {
        Q: 10,
        baseFrequency: 350,
        frequency: 0.5,
        octaves: 3,
        wet: 1
    },
    pingPongDelay: {
        delayTime: 0.25,
        wet: 1
    },
    pitchShift: {
        delayTime: 0,
        feedBack: 0,
        pitch: 0,
        wet: 1,
        windowSize: 0.1
    },
    tremolo: {
        depth: 0.5,
        frequency: 10,
        spread: 180,
        type: 'sine',
        wet: 1
    },
    vibrato: {
        depth: 0.1,
        frequency: 5,
        type: 'sine',
        wet: 1
    }
}