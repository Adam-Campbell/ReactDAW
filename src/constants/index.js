export * from './UIColors';
export * from './programData';
export * from './instrumentData';
export * from './effectData';