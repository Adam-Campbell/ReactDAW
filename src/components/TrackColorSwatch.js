import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { UIColors } from '../constants';
/*



*/

class TrackColorSwatch extends Component {
    constructor(props) {
        super(props);
        this.root = document.getElementById('color-swatch-modal-root');
        this.el = document.createElement('div');
        const { top, left, width } = this.props.nodeRef.getBoundingClientRect();
        this.topPos = top;
        this.leftPos = left + width;
    }

    componentDidMount() {
        this.root.appendChild(this.el);
        window.document.addEventListener('click', this.props.exitColorEditMode);
    }

    componentWillUnmount() {
        this.root.removeChild(this.el);
        window.document.removeEventListener('click', this.props.exitColorEditMode);
    }

    render() {
        return ReactDOM.createPortal(
            <div
                style={{
                    position: 'absolute',
                    top: `${this.topPos}px`,
                    left: `${this.leftPos}px`
                }}
            >
                <span 
                    className="track-color-picker__color-swatch"
                    onClick={() =>this.props.handleColorSwatchClick(UIColors.pink)}
                    style={{
                        backgroundColor: UIColors.pink,
                        color: UIColors.pink
                    }}
                ></span>
                <span 
                    className="track-color-picker__color-swatch"
                    onClick={() =>this.props.handleColorSwatchClick(UIColors.brightBlue)}
                    style={{
                        backgroundColor: UIColors.brightBlue,
                        color: UIColors.brightBlue
                    }}
                ></span>
                <span 
                    className="track-color-picker__color-swatch"
                    onClick={() =>this.props.handleColorSwatchClick(UIColors.brightYellow)}
                    style={{
                        backgroundColor: UIColors.brightYellow,
                        color: UIColors.brightYellow
                    }}
                ></span>
                <span 
                    className="track-color-picker__color-swatch"
                    onClick={() =>this.props.handleColorSwatchClick(UIColors.brightRed)}
                    style={{
                        backgroundColor: UIColors.brightRed,
                        color: UIColors.brightRed
                    }}
                ></span>
                <span 
                    className="track-color-picker__color-swatch"
                    onClick={() =>this.props.handleColorSwatchClick(UIColors.brightGreen)}
                    style={{
                        backgroundColor: UIColors.brightGreen,
                        color: UIColors.brightGreen
                    }}
                ></span>
            </div>,
            this.el
        );
    }
}

export default TrackColorSwatch;