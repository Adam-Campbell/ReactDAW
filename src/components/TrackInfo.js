import React, { Component } from 'react';
import { UIColors } from '../constants';
import { connect } from 'react-redux';
import * as ActionCreators from '../actions';

class TrackInfo extends Component {
    constructor(props) {
        super(props);
        this.foo = 'bar';
        this.inputRef = React.createRef();
        const track = props.channels.find(channel => channel.id === props.trackId);
        this.state = {
            trackName: track.name,
            trackColor: track.color,
            isEditingName: false,
            isEditingTrackColor: false
        };
    }

    updateTrackName = (e) => {
        this.setState({
            trackName: e.target.value
        });
    }

    toggleNameEditing = () => {
        this.setState({
            isEditingName: !this.state.isEditingName
        });
    }

    enterNameEditMode = () => {
        this.setState({
            isEditingName: true
        }, () => {
            this.inputRef.current.focus();
        });
    }

    exitNameEditMode = () => {
        this.setState({
            isEditingName: false
        });
        this.props.updateChannelName(this.props.trackId, this.state.trackName);
    }

    handleInputKeyPress = (e) => {
        if (e.key === 'Enter') {
            this.exitNameEditMode();
        }
    }

    handleMuteButtonClick = (e) => {
        console.log('mute button was clicked');
    }

    handleSoloButtonClick = (e) => {
        console.log('solo button was clicked');
    }

    handleSettingsButtonClick = (e) => {
        console.log('settings button was clicked');
    }

    handleDeleteButtonClick = (e) => {
        this.props.removeChannel(this.props.trackId);
    }

    handleContainerClick = (e) => {
        console.log('container was clicked!');
        if (this.state.isEditingName) {
            this.exitNameEditMode();
        }
    }

    enterColorEditMode = (e) => {
        this.setState({
            isEditingTrackColor: true
        });
    }

    exitColorEditMode = (e) => {
        this.setState({
            isEditingTrackColor: false
        });
    }

    handleColorSwatchClick = (color) => {
        this.setState({
            isEditingTrackColor: false,
            trackColor: color
        });
        this.props.updateChannelColor(this.props.trackId, color);
    }


    render() {
        return (
            <div className="track-info" onClick={this.handleContainerClick}>
                <div className="track-info__left-block">
                    {this.state.isEditingName ?
                        <input 
                            ref={this.inputRef}
                            className="track-info__track-name-input"
                            type="text" 
                            value={this.state.trackName} 
                            onChange={this.updateTrackName}
                            onKeyPress={this.handleInputKeyPress}
                            onClick={(e) => e.stopPropagation()}
                        ></input> :
                        <p 
                            className="track-info__track-name"
                            onClick={this.enterNameEditMode}
                        >{this.state.trackName}</p>  
                    }
                    <button 
                        className="track-info__button" 
                        onClick={this.handleMuteButtonClick}
                    >M</button>
                    <button 
                        className="track-info__button"
                        onClick={this.handleSoloButtonClick}
                    >S</button>
                    <button 
                        className="track-info__button"
                        onClick={this.handleSettingsButtonClick}
                    >O</button>
                    <button 
                        className="track-info__button"
                        onClick={this.handleDeleteButtonClick}
                    >X</button>
                </div>
                <div className="track-color-picker">
                <span 
                    className="track-color-picker__color-swatch"
                    onClick={this.enterColorEditMode}
                    style={{
                        backgroundColor: this.state.trackColor,
                        color: this.state.trackColor
                    }}
                ></span>
                {
                    this.state.isEditingTrackColor && (
                        <React.Fragment>
                            <span 
                                className="track-color-picker__color-swatch"
                                onClick={() =>this.handleColorSwatchClick(UIColors.pink)}
                                style={{
                                    backgroundColor: UIColors.pink,
                                    color: UIColors.pink
                                }}
                            ></span>
                            <span 
                                className="track-color-picker__color-swatch"
                                onClick={() =>this.handleColorSwatchClick(UIColors.brightBlue)}
                                style={{
                                    backgroundColor: UIColors.brightBlue,
                                    color: UIColors.brightBlue
                                }}
                            ></span>
                            <span 
                                className="track-color-picker__color-swatch"
                                onClick={() =>this.handleColorSwatchClick(UIColors.brightYellow)}
                                style={{
                                    backgroundColor: UIColors.brightYellow,
                                    color: UIColors.brightYellow
                                }}
                            ></span>
                            <span 
                                className="track-color-picker__color-swatch"
                                onClick={() =>this.handleColorSwatchClick(UIColors.brightRed)}
                                style={{
                                    backgroundColor: UIColors.brightRed,
                                    color: UIColors.brightRed
                                }}
                            ></span>
                            <span 
                                className="track-color-picker__color-swatch"
                                onClick={() =>this.handleColorSwatchClick(UIColors.brightGreen)}
                                style={{
                                    backgroundColor: UIColors.brightGreen,
                                    color: UIColors.brightGreen
                                }}
                            ></span>
                        </React.Fragment>
                    )
                }
                </div>
            </div>
        );
    }
}

const mapStateToProps = state => ({
    channels: state.channels
});

export default connect(
    mapStateToProps,
    {
        updateChannelName: ActionCreators.updateChannelName,
        updateChannelColor: ActionCreators.updateChannelColor,
        removeChannel: ActionCreators.removeChannel
    }
)(TrackInfo);