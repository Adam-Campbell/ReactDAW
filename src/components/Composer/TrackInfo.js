import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as ActionCreators from '../../actions';
import TrackColorSwatch from './TrackColorSwatch';

class TrackInfo extends Component {
    constructor(props) {
        super(props);
        this.foo = 'bar';
        this.inputRef = React.createRef();
        this.swatchNodeRef = React.createRef();
        const track = props.channels.find(channel => channel.id === props.trackId);
        this.state = {
            trackName: track.name,
            trackColor: track.color,
            isEditingName: false,
            isEditingTrackColor: false
        };
    }

    updateTrackName = (e) => {
        this.setState({
            trackName: e.target.value
        });
    }

    toggleNameEditing = () => {
        this.setState({
            isEditingName: !this.state.isEditingName
        });
    }

    enterNameEditMode = () => {
        this.setState({
            isEditingName: true
        }, () => {
            this.inputRef.current.focus();
        });
    }

    exitNameEditMode = () => {
        this.setState({
            isEditingName: false
        });
        this.props.updateChannelName(this.props.trackId, this.state.trackName);
    }

    handleInputKeyPress = (e) => {
        if (e.key === 'Enter') {
            this.exitNameEditMode();
        }
    }

    handleMuteButtonClick = (e) => {
        //console.log('mute button was clicked');
    }

    handleSoloButtonClick = (e) => {
        //console.log('solo button was clicked');
    }

    handleSettingsButtonClick = (e) => {
        //console.log('settings button was clicked');
        this.openInstrumentWindow();
    }

    handleDeleteButtonClick = (e) => {
        this.props.removeChannel(this.props.trackId);
    }

    handleContainerClick = (e) => {
        //console.log('container was clicked!');
        if (!this.props.isSelected) {
            this.props.updateSelectedChannel(this.props.trackId);
        }
        if (this.state.isEditingName) {
            this.exitNameEditMode();
        }
    }

    enterColorEditMode = (e) => {
        this.setState({
            isEditingTrackColor: true
        });
    }

    exitColorEditMode = (e) => {
        this.setState({
            isEditingTrackColor: false
        });
    }

    handleColorSwatchClick = (color) => {
        this.setState({
            isEditingTrackColor: false,
            trackColor: color
        });
        this.props.updateChannelColor(this.props.trackId, color);
    }

    openInstrumentWindow = () => {
        const track = this.props.channels.find(channel => channel.id === this.props.trackId);
        this.props.openWindow(track.id, 'instrumentSettings');
    }


    render() {
        return (
            <div 
                className={`track-info ${this.props.isSelected && 'selected'}`} 
                onClick={this.handleContainerClick}
            >
                <div className="track-info__left-block">
                    {this.state.isEditingName ?
                        <input 
                            ref={this.inputRef}
                            className="track-info__track-name-input"
                            type="text" 
                            value={this.state.trackName} 
                            onChange={this.updateTrackName}
                            onKeyPress={this.handleInputKeyPress}
                            onClick={(e) => e.stopPropagation()}
                        ></input> :
                        <p 
                            className="track-info__track-name"
                            onClick={this.enterNameEditMode}
                        >{this.state.trackName}</p>  
                    }
                    <button 
                        className="track-info__button" 
                        onClick={this.handleMuteButtonClick}
                    >M</button>
                    <button 
                        className="track-info__button"
                        onClick={this.handleSoloButtonClick}
                    >S</button>
                    <button 
                        className="track-info__button"
                        onClick={this.handleSettingsButtonClick}
                    >O</button>
                    <button 
                        className="track-info__button"
                        onClick={this.handleDeleteButtonClick}
                    >X</button>
                </div>
                <div className="track-color-picker">
                <span 
                    className="track-color-picker__color-swatch"
                    onClick={this.enterColorEditMode}
                    ref={this.swatchNodeRef}
                    style={{
                        backgroundColor: this.state.trackColor,
                        color: this.state.trackColor
                    }}
                ></span>
                {
                    this.state.isEditingTrackColor && (
                        <TrackColorSwatch 
                            handleColorSwatchClick={this.handleColorSwatchClick}
                            nodeRef={this.swatchNodeRef.current}
                            exitColorEditMode={this.exitColorEditMode}
                        />
                    )
                }
                </div>
            </div>
        );
    }
}

TrackInfo.propTypes = {
    trackId: PropTypes.string.isRequired,
    isSelected: PropTypes.bool.isRequired,
    updateSelectedChannel: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
    channels: state.channels
});

export default connect(
    mapStateToProps,
    {
        updateChannelName: ActionCreators.updateChannelName,
        updateChannelColor: ActionCreators.updateChannelColor,
        removeChannel: ActionCreators.removeChannel,
        openWindow: ActionCreators.openWindow
    }
)(TrackInfo);




