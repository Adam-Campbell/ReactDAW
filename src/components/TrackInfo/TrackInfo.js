import React from 'react';
import PropTypes from 'prop-types';
import TrackColorSwatch from './TrackColorSwatch';

const TrackInfo = props => (
    <div 
        className={`track-info ${props.isSelected && 'selected'}`} 
        onClick={props.handleContainerClick}
    >
        <div>
            {props.isEditingName ?
                <input 
                    ref={props.inputRef}
                    className="track-info__track-name-input"
                    type="text" 
                    value={props.trackName} 
                    onChange={props.updateTrackName}
                    onKeyPress={props.handleInputKeyPress}
                    onClick={(e) => e.stopPropagation()}
                ></input> :
                <p 
                    className="track-info__track-name"
                    onClick={props.enterNameEditMode}
                >{props.trackName}</p>  
            }
            <button 
                className={`track-info__button ${props.isMuted ? 'active' : ''}`}
                onClick={props.handleMuteButtonClick}
            >M</button>
            <button 
                className={`track-info__button ${props.isSolo ? 'active' : ''}`}
                onClick={props.handleSoloButtonClick}
            >S</button>
            <button 
                className="track-info__button"
                onClick={props.handleSettingsButtonClick}
            >O</button>
            <button 
                className="track-info__button"
                onClick={props.handleDeleteButtonClick}
            >X</button>
        </div>
        <div className="track-info__color-picker">
            <span 
                className="track-info__color-swatch"
                onClick={props.enterColorEditMode}
                ref={props.swatchNodeRef}
                style={{
                    backgroundColor: props.trackColor,
                    color: props.trackColor
                }}
            ></span>
            {
                props.isEditingTrackColor && (
                    <TrackColorSwatch 
                        handleColorSwatchClick={props.handleColorSwatchClick}
                        nodeRef={props.swatchNodeRef.current}
                        exitColorEditMode={props.exitColorEditMode}
                    />
                )
            }
        </div>
    </div>
);

TrackInfo.propTypes = {
    // react refs
    inputRef: PropTypes.object.isRequired,
    swatchNodeRef: PropTypes.object.isRequired,
    // values
    isSelected: PropTypes.bool.isRequired,
    isEditingName: PropTypes.bool.isRequired,
    trackName: PropTypes.string.isRequired,
    trackColor: PropTypes.string.isRequired,
    isEditingTrackColor: PropTypes.bool.isRequired,
    isMuted: PropTypes.bool.isRequired,
    isSolo: PropTypes.bool.isRequired,
    // callback functions
    handleContainerClick: PropTypes.func.isRequired,
    updateTrackName: PropTypes.func.isRequired,
    handleInputKeyPress: PropTypes.func.isRequired,
    enterNameEditMode: PropTypes.func.isRequired,
    handleMuteButtonClick: PropTypes.func.isRequired,
    handleSoloButtonClick: PropTypes.func.isRequired,
    handleSettingsButtonClick: PropTypes.func.isRequired,
    handleDeleteButtonClick: PropTypes.func.isRequired,
    enterColorEditMode: PropTypes.func.isRequired,
    exitColorEditMode: PropTypes.func.isRequired,
    handleColorSwatchClick: PropTypes.func.isRequired,
};

export default TrackInfo;