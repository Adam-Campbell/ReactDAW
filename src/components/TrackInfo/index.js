import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as ActionCreators from '../../actions';
import TrackInfo from './TrackInfo';

export class TrackInfoContainer extends Component {
    constructor(props) {
        super(props);
        this.inputRef = React.createRef();
        this.swatchNodeRef = React.createRef();
        const track = props.channels.find(channel => channel.id === props.trackId);
        this.state = {
            trackName: track.name,
            trackColor: track.color,
            isEditingName: false,
            isEditingTrackColor: false
        };
    }

    updateTrackName = (e) => {
        this.setState({
            trackName: e.target.value
        });
    }

    toggleNameEditing = () => {
        this.setState({
            isEditingName: !this.state.isEditingName
        });
    }

    enterNameEditMode = () => {
        this.setState({
            isEditingName: true
        }, () => {
            this.inputRef.current.focus();
        });
    }

    exitNameEditMode = () => {
        this.setState({
            isEditingName: false
        });
        this.props.updateChannelName(this.props.trackId, this.state.trackName);
    }

    handleInputKeyPress = (e) => {
        if (e.key === 'Enter') {
            this.exitNameEditMode();
        }
    }

    handleMuteButtonClick = (e) => {
        //console.log('mute button was clicked');
    }

    handleSoloButtonClick = (e) => {
        //console.log('solo button was clicked');
    }

    handleSettingsButtonClick = (e) => {
        //console.log('settings button was clicked');
        this.openInstrumentWindow();
    }

    handleDeleteButtonClick = (e) => {
        this.props.removeChannel(this.props.trackId);
    }

    handleContainerClick = (e) => {
        //console.log('container was clicked!');
        if (!this.props.isSelected) {
            this.props.updateSelectedChannel(this.props.trackId);
        }
        if (this.state.isEditingName) {
            this.exitNameEditMode();
        }
    }

    enterColorEditMode = (e) => {
        this.setState({
            isEditingTrackColor: true
        });
    }

    exitColorEditMode = (e) => {
        this.setState({
            isEditingTrackColor: false
        });
    }

    handleColorSwatchClick = (color) => {
        this.setState({
            isEditingTrackColor: false,
            trackColor: color
        });
        this.props.updateChannelColor(this.props.trackId, color);
    }

    openInstrumentWindow = () => {
        const track = this.props.channels.find(channel => channel.id === this.props.trackId);
        this.props.openWindow(track.id, 'instrumentSettings');
    }


    render() {
        return <TrackInfo 
            inputRef={this.inputRef}
            swatchNodeRef={this.swatchNodeRef}
            isSelected={this.props.isSelected}
            isEditingName={this.state.isEditingName}
            trackName={this.state.trackName}
            isEditingTrackColor={this.state.isEditingTrackColor}
            trackColor={this.state.trackColor}
            handleContainerClick={this.handleContainerClick}
            updateTrackName={this.updateTrackName}
            handleInputKeyPress={this.handleInputKeyPress}
            enterNameEditMode={this.enterNameEditMode}
            handleMuteButtonClick={this.handleMuteButtonClick}
            handleSoloButtonClick={this.handleSoloButtonClick}
            handleSettingsButtonClick={this.handleSettingsButtonClick}
            handleDeleteButtonClick={this.handleDeleteButtonClick}
            enterColorEditMode={this.enterColorEditMode}
            exitColorEditMode={this.exitColorEditMode}
            handleColorSwatchClick={this.handleColorSwatchClick}
        />
    }
}

TrackInfoContainer.propTypes = {
    trackId: PropTypes.string.isRequired,
    isSelected: PropTypes.bool.isRequired,
    updateSelectedChannel: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
    channels: state.channels
});

export default connect(
    mapStateToProps,
    {
        updateChannelName: ActionCreators.updateChannelName,
        updateChannelColor: ActionCreators.updateChannelColor,
        removeChannel: ActionCreators.removeChannel,
        openWindow: ActionCreators.openWindow
    }
)(TrackInfoContainer);




