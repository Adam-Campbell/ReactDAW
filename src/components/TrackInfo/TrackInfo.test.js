import React from 'react';
import { TrackInfoContainer } from './';
import TrackInfo from './TrackInfo';
import TrackColorSwatch from './TrackColorSwatch';
import { shallow } from 'enzyme';
import { UIColors } from '../../constants';

const channelsState = [
    {
      id: '2481615053298156',
      name: 'Channel 1',
      color: '#d86597',
      instrumentId: '1496404507349548',
      effectIds: [],
      sectionIds: [],
      isMuted: false,
      isSolo: false,
      volume: 0,
      pan: 0
    },
    {
      id: '0271578528326165',
      name: 'Channel 2',
      color: '#d86597',
      instrumentId: '0446480036600608',
      effectIds: [],
      sectionIds: [],
      isMuted: false,
      isSolo: false,
      volume: 0,
      pan: 0
    },
    {
      id: '9060864310622598',
      name: 'Channel 3',
      color: '#d86597',
      instrumentId: '8229538223456731',
      effectIds: [],
      sectionIds: [],
      isMuted: false,
      isSolo: false,
      volume: 0,
      pan: 0
    }
];

test('container component renders correctly', () => {
    const component = shallow(
        <TrackInfoContainer 
           trackId="2481615053298156"
           isSelected={false}
           updateSelectedChannel={jest.fn()}
           channels={channelsState}
           updateChannelName={jest.fn()}
           updateChannelColor={jest.fn()}
           removeChannel={jest.fn()}
           openWindow={jest.fn()} 
        />
    );
    expect(component).toMatchSnapshot();
});

test('presentational component renders correctly', () => {
    const component = shallow(
        <TrackInfo 
            inputRef={React.createRef()}
            swatchNodeRef={React.createRef()}
            isSelected={false}
            isEditingName={false}
            isEditingTrackColor={false}
            isMuted={false}
            isSolo={false}
            trackName="Channel 1"
            trackColor={UIColors.pink}
            handleContainerClick={jest.fn()}
            updateTrackName={jest.fn()}
            handleInputKeyPress={jest.fn()}
            enterNameEditMode={jest.fn()}
            handleMuteButtonClick={jest.fn()}
            handleSoloButtonClick={jest.fn()}
            handleSettingsButtonClick={jest.fn()}
            handleDeleteButtonClick={jest.fn()}
            enterColorEditMode={jest.fn()}
            exitColorEditMode={jest.fn()}
            handleColorSwatchClick={jest.fn()}
        />
    );
    expect(component).toMatchSnapshot();
});
