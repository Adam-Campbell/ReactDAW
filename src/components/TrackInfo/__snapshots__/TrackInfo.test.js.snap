// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container component renders correctly 1`] = `
<TrackInfo
  enterColorEditMode={[Function]}
  enterNameEditMode={[Function]}
  exitColorEditMode={[Function]}
  handleColorSwatchClick={[Function]}
  handleContainerClick={[Function]}
  handleDeleteButtonClick={[Function]}
  handleInputKeyPress={[Function]}
  handleMuteButtonClick={[Function]}
  handleSettingsButtonClick={[Function]}
  handleSoloButtonClick={[Function]}
  inputRef={
    Object {
      "current": null,
    }
  }
  isEditingName={false}
  isEditingTrackColor={false}
  isMuted={false}
  isSelected={false}
  isSolo={false}
  swatchNodeRef={
    Object {
      "current": null,
    }
  }
  trackColor="#d86597"
  trackName="Channel 1"
  updateTrackName={[Function]}
/>
`;

exports[`presentational component renders correctly 1`] = `
<div
  className="track-info false"
  onClick={[MockFunction]}
>
  <div
    className="track-info__left-block"
  >
    <p
      className="track-info__track-name"
      onClick={[MockFunction]}
    >
      Channel 1
    </p>
    <button
      className="track-info__button "
      onClick={[MockFunction]}
    >
      M
    </button>
    <button
      className="track-info__button "
      onClick={[MockFunction]}
    >
      S
    </button>
    <button
      className="track-info__button"
      onClick={[MockFunction]}
    >
      O
    </button>
    <button
      className="track-info__button"
      onClick={[MockFunction]}
    >
      X
    </button>
  </div>
  <div
    className="track-color-picker"
  >
    <span
      className="track-color-picker__color-swatch"
      onClick={[MockFunction]}
      style={
        Object {
          "backgroundColor": "#d86597",
          "color": "#d86597",
        }
      }
    />
  </div>
</div>
`;
